e2e/stack_test.go

```diff
@@ -282,48 +272,244 @@ func depositE2E(t *testing.T, stack *e2e.StackConfig) {
 	query := fmt.Sprintf(
-		"%s.%s='%s'",
-		rolluptypes.EventTypeMintETH,
-		rolluptypes.AttributeKeyL1DepositTxType,
-		rolluptypes.L1UserDepositTxType,
+		"%s.%s='%s' AND %s.%s='%s' AND %s.%s='%s'",
+		rolluptypes.EventTypeMintETH, rolluptypes.AttributeKeyL1DepositTxType, rolluptypes.L1UserDepositTxType,
+		rolluptypes.EventTypeMintETH, rolluptypes.AttributeKeyToCosmosAddress, userAddress,
+		rolluptypes.EventTypeMintETH, rolluptypes.AttributeKeyValue, valueHex,
 	)
+	result := requireL2Event(t, stack, query)
+
+	require.NotEmpty(t, result.Txs, "mint_eth event not found")
+	t.Log("Monomer can mint ETH for L1 user deposits")
+}
+
+func requireEthIsBurnedAndGetL2BlockNum(t *testing.T, stack *e2e.StackConfig, userAddress, valueHex string) *big.Int {
+	query := fmt.Sprintf(
+		"%s.%s='%s' AND %s.%s='%s' AND %s.%s='%s'",
+		rolluptypes.EventTypeBurnETH, rolluptypes.AttributeKeyL2WithdrawalTx, rolluptypes.EventTypeWithdrawalInitiated,
+		rolluptypes.EventTypeBurnETH, rolluptypes.AttributeKeyFromCosmosAddress, userAddress,
+		rolluptypes.EventTypeBurnETH, rolluptypes.AttributeKeyValue, valueHex,
+	)
+	result := requireL2Event(t, stack, query)
+
+	require.NotEmpty(t, result.Txs, "burn_eth event not found")
+	t.Log("Monomer can burn ETH for L2 user withdrawals")
+
+	return big.NewInt(result.Txs[0].Height)
+}
+
+func requireL2Event(t *testing.T, stack *e2e.StackConfig, query string) *cometcore.ResultTxSearch {
 	page := 1
 	perPage := 100
-	orderBy := "desc"
 
-	result, err := appchainClient.TxSearch(
-		context.Background(),
+	result, err := stack.L2Client.TxSearch(
+		stack.Ctx,
 		query,
 		false,
 		&page,
 		&perPage,
-		orderBy,
+		"desc",
 	)
 	require.NoError(t, err, "search transactions")
 	require.NotNil(t, result)
-	require.NotEmpty(t, result.Txs, "mint_eth event not found")
-	t.Log("Monomer can mint_eth from L1 user deposits")
+	return result
+}
+
+func createL1TransactOpts(
+	t *testing.T,
+	stack *e2e.StackConfig,
+	user *e2e.L1User,
+	l1signer types.Signer,
+	l1GasLimit uint64,
+	value *big.Int,
+) *bind.TransactOpts {
+	return &bind.TransactOpts{
+		From: user.Address,
+		Signer: func(addr common.Address, tx *types.Transaction) (*types.Transaction, error) {
+			signed, err := types.SignTx(tx, l1signer, user.PrivateKey)
+			if err != nil {
+				return nil, err
+			}
+			return signed, nil
+		},
+		Nonce:    getCurrentUserNonce(t, stack, user.Address),
+		GasPrice: getSuggestedL1GasPrice(t, stack),
+		GasLimit: l1GasLimit,
+		Value:    value,
+		Context:  stack.Ctx,
+		NoSend:   false,
+	}
+}
+
+func getCurrentUserNonce(t *testing.T, stack *e2e.StackConfig, userAddress common.Address) *big.Int {
+	nonce, err := stack.L1Client.NonceAt(stack.Ctx, userAddress, nil)
+	require.NoError(t, err)
+	return new(big.Int).SetUint64(nonce)
+}
+
+func getSuggestedL1GasPrice(t *testing.T, stack *e2e.StackConfig) *big.Int {
+	gasPrice, err := stack.L1Client.Client.SuggestGasPrice(stack.Ctx)
+	require.NoError(t, err)
+	return gasPrice
+}
```

Comment:
I really enjoyed this comment and I find it compelling.

Counterpoint: I mostly think of one-off functions as compiler-enforced comments. Adding an inline `// doing thing` before a handful of crufty code
- has legibility problems, because the code "in scope" of that comment is not obvious
- carries liabilities into the future, because the in-scope code is as likely as not to get mixed with other code, or maybe `thing` doesn't even get done later but the comment remains.

If, instead, you define and call `doThing()`, it's
- obvious when `thing` is done
- easier, at least on subsequent reads to gloss directly over top of the invocation of doThing
- less susceptible to drift between the comment (function name) and implementation


I don't know why this is true, but for some reason I'm not *upset* by package namespace clutter in the same way that I am about interfaces or classes. Maybe because I don't use tooling to discover these package namespaced items?

Reply:
Fair.

I don't feel strongly; I'll resolve this to avoid bikeshedding, which would definitely be the worst outcome.